src/backend/distributed/connection/locally_reserved_shared_connection_stats.c

```diff
@@ -0,0 +1,487 @@
+						errhint("Set citus.max_shared_pool_size TO -1 to "
+								"disable reserved connection counters")));
+	}
+
+	entry->reservedConnectionCount--;
+
+	ereport(DEBUG5, (errmsg("Decrement shared reserved connection to node %s:%d to %d",
+							hostName, nodePort, entry->reservedConnectionCount)));
+}
+
+
+/*
+ * ReserveSharedConnectionCounterForAllPrimaryNodesIfNeeded is a wrapper around
+ * ReserveSharedConnectionCounterForNodeListIfNeeded.
+ */
+void
+ReserveSharedConnectionCounterForAllPrimaryNodesIfNeeded(void)
+{
+	List *primaryNodeList = ActivePrimaryNonCoordinatorNodeList(NoLock);
+
+	ReserveSharedConnectionCounterForNodeListIfNeeded(primaryNodeList);
+}
+
+
+/*
+ * ReserveSharedConnectionCounterForNodeListIfNeeded reserves a shared connection
+ * counter per node in the nodeList unless:
+ *  - there is at least one connection to the node
+ *  - has already reserved a connection to the node
+ */
+static void
+ReserveSharedConnectionCounterForNodeListIfNeeded(List *nodeList)
+{
+	if (GetMaxSharedPoolSize() == DISABLE_CONNECTION_THROTTLING)
+	{
+		/* connection throttling disabled */
+		return;
+	}
+
+	if (SessionLocalReservedConnectionCounters == NULL)
+	{
+		/*
+		 * This is unexpected as SessionLocalReservedConnectionCounters hash table is
+		 * created at startup. Still, let's be defensive.
+		 */
+		ereport(ERROR, (errmsg("unexpected state for session level reserved "
+							   "connection counter hash")));
+	}
+
+	/*
+	 * We sort the workerList because adaptive connection management
+	 * (e.g., OPTIONAL_CONNECTION) requires any concurrent executions
+	 * to wait for the connections in the same order to prevent any
+	 * starvation. If we don't sort, we might end up with:
+	 *      Execution 1: Get connection for worker 1, wait for worker 2
+	 *      Execution 2: Get connection for worker 2, wait for worker 1
+	 *
+	 *  and, none could proceed. Instead, we enforce every execution establish
+	 *  the required connections to workers in the same order.
+	 */
+	nodeList = SortList(nodeList, CompareWorkerNodes);
+
+	char *databaseName = get_database_name(MyDatabaseId);
+
+	WorkerNode *workerNode = NULL;
+	foreach_ptr(workerNode, nodeList)
+	{
+		if (ConnectionAvaliableToNode(workerNode->workerName, workerNode->workerPort,
+									  CurrentUserName(), databaseName))
+		{
+			/*
+			 * The same user has already an active connection for the node. It
+			 * means that the execution can use the same connection, so reservation
+			 * is not necessary.
+			 */
+			continue;
+		}
+
+		if (HasAlreadyReservedConnection(workerNode->workerName, workerNode->workerPort,
+										 MyDatabaseId))
+		{
+			/*
+			 * We already locally reserved a connection for this node. So, it is not
+			 * necessary to reserve again.
+			 *
+			 * One caveat here is that we are ignoring for which user we reserved the
+			 * connection. It is not important as Citus doesn't support switching
+			 * between users (https://github.com/citusdata/citus/pull/3869) and
```

Comment:
It does, just not when we need to read over a particular connection that did a modification as a different user. Is that still ok?

Reply:
Add user name as well

Reply:
I'm not sure we should remove this check. If we keep this check, nested calls will not do a new reservation if one already exists, so we might end up in a situation where a nested call uses our reservation, but in that case it will leave behind a connection that we can use. If we then use that connection and another nested call happens, it will do a new reservation because there is none yet.